pg_search/src/query/builder.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2023-2025 ParadeDB, Inc.
+//
+// This file is part of ParadeDB - Postgres for Search and Analytics
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+//! Generic query builder trait for different output types.
+
+use super::estimate_tree::QueryWithEstimates;
+use super::SearchQueryInput;
+use tantivy::query::Query as TantivyQuery;
+
+/// Build different output types from Tantivy queries.
+pub trait QueryBuilder {
+    type Output;
+
+    /// Build a leaf node with no children.
+    fn build_leaf(
+        &self,
+        query_input: SearchQueryInput,
+        tantivy_query: Box<dyn TantivyQuery>,
+        label: String,
+    ) -> Self::Output;
+
+    /// Build a node with children.
+    fn build_with_children(
+        &self,
+        query_input: SearchQueryInput,
+        tantivy_query: Box<dyn TantivyQuery>,
+        label: String,
+        children: Vec<Self::Output>,
+    ) -> Self::Output;
+
+    /// Extract the query from output.
+    fn extract_query(output: &Self::Output) -> &dyn TantivyQuery;
+
+    /// Clone the query from output.
+    fn clone_query(output: &Self::Output) -> Box<dyn TantivyQuery>;
+}
+
+/// Builder that returns only the Tantivy Query.
+pub struct QueryOnlyBuilder;
+
+impl QueryBuilder for QueryOnlyBuilder {
+    type Output = Box<dyn TantivyQuery>;
+
+    fn build_leaf(
+        &self,
+        _query_input: SearchQueryInput,
+        tantivy_query: Box<dyn TantivyQuery>,
+        _label: String,
+    ) -> Self::Output {
+        tantivy_query
+    }
+
+    fn build_with_children(
+        &self,
+        _query_input: SearchQueryInput,
+        tantivy_query: Box<dyn TantivyQuery>,
+        _label: String,
+        _children: Vec<Self::Output>,
+    ) -> Self::Output {
+        tantivy_query
+    }
+
+    fn extract_query(output: &Self::Output) -> &dyn TantivyQuery {
+        output.as_ref()
+    }
+
+    fn clone_query(output: &Self::Output) -> Box<dyn TantivyQuery> {
+        output.box_clone()
+    }
+}
+
+/// Builder that returns both Tantivy Query and QueryWithEstimates tree.
+pub struct QueryTreeBuilder;
+
+impl QueryBuilder for QueryTreeBuilder {
+    type Output = (Box<dyn TantivyQuery>, QueryWithEstimates);
+
+    fn build_leaf(
+        &self,
+        query_input: SearchQueryInput,
```

Comment:
Can this be a reference?
```suggestion
        query_input: &SearchQueryInput,
```

AFAICT, `into_tantivy_query_generic` is passing in roughly a copy of the node that it is operating on, and it looks like it could pass a reference to it instead?